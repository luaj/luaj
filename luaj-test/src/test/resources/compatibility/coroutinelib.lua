function printrunning()
	if coroutine.running() == nil then 
		print("running is nil");
	else
		print("running is not nil")
	end
end

function foo (a)
	print("foo", a)
    return coroutine.yield(2*a)
end

co = coroutine.create(function (a,b)
	print("co-body", a, b)
	local r = foo(a+1)
	print("co-body", r)
	local r, s = coroutine.yield(a+b, a-b)
	print("co-body", r, s)

	printrunning()
	print("co.status.inside",coroutine.status(co));
	local co2 = coroutine.create(function()
		print("co.status.inside2",coroutine.status(co));
	end)
	print("co.status.inside",coroutine.status(co));
	coroutine.resume(co2);
		
	return b, "end"
end)

function exercise() 
	printrunning()
	print("co.status",coroutine.status(co));
	print("main", coroutine.resume(co, 1, 10))
	print("co.status",coroutine.status(co));
	print("main", coroutine.resume(co, "r"))
	print("co.status",coroutine.status(co));
	print("main", coroutine.resume(co, "x", "y"))
	print("co.status",coroutine.status(co));
	print("main", coroutine.resume(co, "x", "y"))
	print("co.status",coroutine.status(co));
end

exercise();

co = coroutine.create(function (a,b)
	print("co-body", a, b)
-- TODO: make java and C behave the same for yielding in pcalls
--	local status,r = pcall( foo, a+1 )
foo(a+1)
	print("co-body", status,r)
	local r, s = coroutine.yield(a+b, a-b)
	print("co-body", r, s)
	return b, "end"
end)

exercise();


-- wrap test 
local g = coroutine.wrap(function (a,b)
	print("co-body", a, b)
	local r = foo(a+1)
	print("co-body", r)
	local r, s = coroutine.yield(a+b, a-b)
	print("co-body", r, s)
	return b, "end"
end )

print("g", g(1, 10))
print("g", g("r"))
print("g", g("x", "y"))
local s,e = pcall( g, "x", "y" )
print("g", string.match(e,'cannot resume dead coroutine') or 'badmessage: '..tostring(e))

-- varargs passing
local echo = function(msg,...)
	print( msg, ...)
	return ... 
end
local echocr = function(...)
	local arg = table.pack(...) 
	echo('(echocr) first args', table.unpack(arg,1,arg.n)) 
	local a = arg
	while true do
		a = { echo( '(echoch) yield returns', coroutine.yield( table.unpack(a) ) ) }
	end
end
local c = coroutine.create( echocr )
local step = function(...)
	echo( '(main) resume returns', 
		coroutine.resume(c, echo('(main) sending args', ...)) ) 
end
step(111,222,333)
step()
step(111)
step(111,222,333)

-- test loops in resume calls
b = coroutine.create( function( arg )
	while ( true ) do
		print( '  b-resumed', arg, b == coroutine.running() )
		print( '  b-b', coroutine.status(b) ) 
		print( '  b-c', coroutine.status(c) ) 
		print( '  b-resume-b',coroutine.resume( b, 'b-arg-for-b' ) )
		print( '  b-resume-c',coroutine.resume( c, 'b-arg-for-c' ) )
		arg = coroutine.yield( 'b-rslt' )
	end
end )
c = coroutine.create( function( arg )
	for i=1,3 do
		print( '    c-resumed', arg, c == coroutine.running() )
		print( '    c-b', coroutine.status(b) ) 
		print( '    c-c', coroutine.status(c) ) 
		print( '    c-resume-b',coroutine.resume( b, 'b-arg-for-b' ) )
		print( '    c-resume-c',coroutine.resume( c, 'b-arg-for-c' ) )
		arg = coroutine.yield( 'c-rslt' )
	end
end )
for i=1,3 do
	print( 'main-b', coroutine.status(b) )
	print( 'main-c', coroutine.status(c) )
	print( 'main-resume-b',coroutine.resume( b, 'main-arg-for-b' ) )
	print( 'main-resume-c',coroutine.resume( c, 'main-arg-for-c' ) )
end
