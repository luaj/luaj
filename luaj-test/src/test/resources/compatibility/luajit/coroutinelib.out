running is nil
co.status	suspended
co-body	1	10
foo	2
main	true	4
co.status	suspended
co-body	r
main	true	11	-9
co.status	suspended
co-body	x	y
running is not nil
co.status.inside	running
co.status.inside	running
co.status.inside2	normal
main	true	10	end
co.status	dead
main	false	cannot resume dead coroutine
co.status	dead
running is nil
co.status	suspended
co-body	1	10
foo	2
main	true	4
co.status	suspended
co-body	nil	nil
main	true	11	-9
co.status	suspended
co-body	x	y
main	true	10	end
co.status	dead
main	false	cannot resume dead coroutine
co.status	dead
co-body	1	10
foo	2
g	4
co-body	r
g	11	-9
co-body	x	y
g	10	end
g	cannot resume dead coroutine
(main) sending args	111	222	333
(main) resume returns	false	coroutinelib.lua:83: attempt to call field 'pack' (a nil value)
(main) sending args
(main) resume returns	false	cannot resume dead coroutine
(main) sending args	111
(main) resume returns	false	cannot resume dead coroutine
(main) sending args	111	222	333
(main) resume returns	false	cannot resume dead coroutine
main-b	suspended
main-c	suspended
  b-resumed	main-arg-for-b	true
  b-b	running
  b-c	suspended
  b-resume-b	false	cannot resume running coroutine
    c-resumed	b-arg-for-c	true
    c-b	normal
    c-c	running
    c-resume-b	false	cannot resume running coroutine
    c-resume-c	false	cannot resume running coroutine
  b-resume-c	true	c-rslt
main-resume-b	true	b-rslt
    c-resumed	main-arg-for-c	true
    c-b	suspended
    c-c	running
  b-resumed	b-arg-for-b	true
  b-b	running
  b-c	normal
  b-resume-b	false	cannot resume running coroutine
  b-resume-c	false	cannot resume running coroutine
    c-resume-b	true	b-rslt
    c-resume-c	false	cannot resume running coroutine
main-resume-c	true	c-rslt
main-b	suspended
main-c	suspended
  b-resumed	main-arg-for-b	true
  b-b	running
  b-c	suspended
  b-resume-b	false	cannot resume running coroutine
    c-resumed	b-arg-for-c	true
    c-b	normal
    c-c	running
    c-resume-b	false	cannot resume running coroutine
    c-resume-c	false	cannot resume running coroutine
  b-resume-c	true	c-rslt
main-resume-b	true	b-rslt
main-resume-c	true
main-b	suspended
main-c	dead
  b-resumed	main-arg-for-b	true
  b-b	running
  b-c	dead
  b-resume-b	false	cannot resume running coroutine
  b-resume-c	false	cannot resume dead coroutine
main-resume-b	true	b-rslt
main-resume-c	false	cannot resume dead coroutine
